# Семинар 2. Исследование методов линейной регрессии

#### Критерий оценивания:
#### Пункт 14: максимум 0,5 балла
#### Пункт 15: максимум 0,5 балла

#### Итого за работу: максимум 1 балл
#### P.S. пункты 14 и 15 без пунктов 1-13 не засчитываются, ответы на вопросы из ИИ не засчитываются

1. Загрузите датасет и выведите на экран первые несколько строк

In [1]:
import numpy as np
import pandas as pd
from IPython.display import display

RANDOM_STATE = 42

data = pd.read_csv('auto_dataset.csv')
data.head()


,brand,model,vehicleType,gearbox,fuelType,notRepairedDamage,powerPS,kilometer,autoAgeMonths,price
0,volkswagen,golf,kleinwagen,manuell,benzin,nein,75,150000,177,1500
1,skoda,fabia,kleinwagen,manuell,diesel,nein,69,90000,93,3600
2,bmw,3er,limousine,manuell,benzin,ja,102,150000,246,650
3,peugeot,2_reihe,cabrio,manuell,benzin,nein,109,150000,140,2200
4,mazda,3_reihe,limousine,manuell,benzin,nein,105,150000,136,2000


2. Разбейте выборку на признаки и ответы. Закодируйте категориальные признаки.

In [2]:
X = data.drop(columns='price')
y = data['price'].astype(float)

categorical_columns = X.select_dtypes(include='object').columns
X = pd.get_dummies(X, columns=categorical_columns, drop_first=True, dtype=float)

print('Размер X:', X.shape)
print('Размер y:', y.shape)
X.head()


Размер X: (1000, 204)
Размер y: (1000,)


,powerPS,kilometer,autoAgeMonths,brand_audi,brand_bmw,brand_chevrolet,brand_chrysler,brand_citroen,brand_dacia,brand_daewoo,brand_daihatsu,brand_fiat,brand_ford,brand_honda,brand_hyundai,brand_jaguar,brand_jeep,brand_kia,brand_lancia,brand_land_rover,brand_mazda,brand_mercedes_benz,brand_mini,brand_mitsubishi,brand_nissan,brand_opel,brand_peugeot,brand_porsche,brand_renault,brand_rover,brand_saab,brand_seat,brand_skoda,brand_smart,brand_subaru,brand_suzuki,brand_toyota,brand_volkswagen,brand_volvo,model_156,...,model_sl,model_slk,model_sorento,model_sportage,model_sprinter,model_stilo,model_swift,model_tiguan,model_touareg,model_touran,model_transporter,model_tt,model_twingo,model_up,model_v40,model_v50,model_v_klasse,model_vectra,model_verso,model_viano,model_vito,model_x_reihe,model_yaris,model_yeti,model_ypsilon,model_z_reihe,model_zafira,vehicleType_bus,vehicleType_cabrio,vehicleType_coupe,vehicleType_kleinwagen,vehicleType_kombi,vehicleType_limousine,vehicleType_suv,gearbox_manuell,fuelType_benzin,fuelType_diesel,fuelType_hybrid,fuelType_lpg,notRepairedDamage_nein
0,75,150000,177,0.0,0.0,0.0,0.0,0.0,0.0,0.0,0.0,0.0,0.0,0.0,0.0,0.0,0.0,0.0,0.0,0.0,0.0,0.0,0.0,0.0,0.0,0.0,0.0,0.0,0.0,0.0,0.0,0.0,0.0,0.0,0.0,0.0,0.0,1.0,0.0,0.0,...,0.0,0.0,0.0,0.0,0.0,0.0,0.0,0.0,0.0,0.0,0.0,0.0,0.0,0.0,0.0,0.0,0.0,0.0,0.0,0.0,0.0,0.0,0.0,0.0,0.0,0.0,0.0,0.0,0.0,0.0,1.0,0.0,0.0,0.0,1.0,1.0,0.0,0.0,0.0,1.0
1,69,90000,93,0.0,0.0,0.0,0.0,0.0,0.0,0.0,0.0,0.0,0.0,0.0,0.0,0.0,0.0,0.0,0.0,0.0,0.0,0.0,0.0,0.0,0.0,0.0,0.0,0.0,0.0,0.0,0.0,0.0,1.0,0.0,0.0,0.0,0.0,0.0,0.0,0.0,...,0.0,0.0,0.0,0.0,0.0,0.0,0.0,0.0,0.0,0.0,0.0,0.0,0.0,0.0,0.0,0.0,0.0,0.0,0.0,0.0,0.0,0.0,0.0,0.0,0.0,0.0,0.0,0.0,0.0,0.0,1.0,0.0,0.0,0.0,1.0,0.0,1.0,0.0,0.0,1.0
2,102,150000,246,0.0,1.0,0.0,0.0,0.0,0.0,0.0,0.0,0.0,0.0,0.0,0.0,0.0,0.0,0.0,0.0,0.0,0.0,0.0,0.0,0.0,0.0,0.0,0.0,0.0,0.0,0.0,0.0,0.0,0.0,0.0,0.0,0.0,0.0,0.0,0.0,0.0,...,0.0,0.0,0.0,0.0,0.0,0.0,0.0,0.0,0.0,0.0,0.0,0.0,0.0,0.0,0.0,0.0,0.0,0.0,0.0,0.0,0.0,0.0,0.0,0.0,0.0,0.0,0.0,0.0,0.0,0.0,0.0,0.0,1.0,0.0,1.0,1.0,0.0,0.0,0.0,0.0
3,109,150000,140,0.0,0.0,0.0,0.0,0.0,0.0,0.0,0.0,0.0,0.0,0.0,0.0,0.0,0.0,0.0,0.0,0.0,0.0,0.0,0.0,0.0,0.0,0.0,1.0,0.0,0.0,0.0,0.0,0.0,0.0,0.0,0.0,0.0,0.0,0.0,0.0,0.0,...,0.0,0.0,0.0,0.0,0.0,0.0,0.0,0.0,0.0,0.0,0.0,0.0,0.0,0.0,0.0,0.0,0.0,0.0,0.0,0.0,0.0,0.0,0.0,0.0,0.0,0.0,0.0,0.0,1.0,0.0,0.0,0.0,0.0,0.0,1.0,1.0,0.0,0.0,0.0,1.0
4,105,150000,136,0.0,0.0,0.0,0.0,0.0,0.0,0.0,0.0,0.0,0.0,0.0,0.0,0.0,0.0,0.0,0.0,0.0,1.0,0.0,0.0,0.0,0.0,0.0,0.0,0.0,0.0,0.0,0.0,0.0,0.0,0.0,0.0,0.0,0.0,0.0,0.0,0.0,...,0.0,0.0,0.0,0.0,0.0,0.0,0.0,0.0,0.0,0.0,0.0,0.0,0.0,0.0,0.0,0.0,0.0,0.0,0.0,0.0,0.0,0.0,0.0,0.0,0.0,0.0,0.0,0.0,0.0,0.0,0.0,0.0,1.0,0.0,1.0,1.0,0.0,0.0,0.0,1.0


3. Разбейте датасет на train val test в отношении 8:1:1

In [3]:
# Разбиение 8:1:1. Все статистики масштабирования считаются только по train.
rng = np.random.default_rng(RANDOM_STATE)
indices = rng.permutation(len(X))

n_train = int(0.8 * len(X))
n_val = int(0.1 * len(X))

train_idx = indices[:n_train]
val_idx = indices[n_train:n_train + n_val]
test_idx = indices[n_train + n_val:]

X_np = X.to_numpy(dtype=float)
y_np = y.to_numpy(dtype=float)

X_train = X_np[train_idx]
X_val = X_np[val_idx]
X_test = X_np[test_idx]
y_train = y_np[train_idx]
y_val = y_np[val_idx]
y_test = y_np[test_idx]

# Масштабирование признаков нужно для устойчивой работы градиентных методов.
mean_train = X_train.mean(axis=0)
std_train = X_train.std(axis=0)
std_train[std_train == 0] = 1.0

X_train = (X_train - mean_train) / std_train
X_val = (X_val - mean_train) / std_train
X_test = (X_test - mean_train) / std_train

# Добавляем свободный коэффициент (bias).
X_train = np.column_stack([np.ones(len(X_train)), X_train])
X_val = np.column_stack([np.ones(len(X_val)), X_val])
X_test = np.column_stack([np.ones(len(X_test)), X_test])

print('train:', X_train.shape, y_train.shape)
print('val:  ', X_val.shape, y_val.shape)
print('test: ', X_test.shape, y_test.shape)


train: (800, 205) (800,)
val:   (100, 205) (100,)
test:  (100, 205) (100,)


4. Исследуйте VGD с постоянным шагом n:

переберите n в логарифмической сетке от 10^-5 до 1

для каждого n:

* обучите VGD на train
* найдите и запомните R^2_train и Loss_train
* протестируйте VGD на val, запомните Loss_val

найдите наилучший n по минимальному Loss_val, запомните его Loss_train, R^2_train, Loss_val;

протестируйте VGD c лучшим n на test, запомните Loss_test, R^2_test, число итераций на test.

In [4]:
def mse(X, y, w):
    residuals = X @ w - y
    return float(np.mean(residuals ** 2))


def r2_score_manual(X, y, w):
    prediction = X @ w
    ss_res = np.sum((y - prediction) ** 2)
    ss_tot = np.sum((y - y.mean()) ** 2)
    return float(1.0 - ss_res / ss_tot)


def fit_optimizer(method, learning_rate, use_time_decay=False,
                  max_iter=10_000, tol=1e-4, batch_size=32,
                  random_state=RANDOM_STATE):
    """Обучение одного из пяти методов линейной регрессии."""
    n, d = X_train.shape
    w = np.zeros(d, dtype=float)
    rng = np.random.default_rng(random_state)

    # Для full-batch методов заранее считаем матрицы градиента MSE.
    if method in {'VGD', 'Momentum', 'Adam'}:
        a = (2.0 / n) * (X_train.T @ X_train)
        b = (2.0 / n) * (X_train.T @ y_train)

    if method == 'Momentum':
        h = np.zeros(d, dtype=float)

    if method == 'Adam':
        m = np.zeros(d, dtype=float)
        v = np.zeros(d, dtype=float)

    if method == 'SAG':
        gradient_memory = np.zeros((n, d), dtype=float)
        mean_gradient = np.zeros(d, dtype=float)

    previous_loss = None
    stable_checks = 0
    check_every = 20

    with np.errstate(over='ignore', invalid='ignore', divide='ignore'):
        for k in range(max_iter):
            eta = float(learning_rate)
            if use_time_decay:
                # TimeDecayLR из лекции: eta_k = lambda * (s0 / (s0 + k))^p,
                # где s0 = 1, p = 0.5.
                eta *= (1.0 / (1.0 + k)) ** 0.5

            if method == 'VGD':
                gradient = a @ w - b
                w -= eta * gradient

            elif method == 'SGD':
                batch = rng.choice(n, size=min(batch_size, n), replace=False)
                X_batch = X_train[batch]
                y_batch = y_train[batch]
                gradient = (2.0 / len(batch)) * X_batch.T @ (X_batch @ w - y_batch)
                w -= eta * gradient

            elif method == 'SAG':
                j = int(rng.integers(n))
                x_j = X_train[j]
                new_gradient = 2.0 * (x_j @ w - y_train[j]) * x_j
                mean_gradient += (new_gradient - gradient_memory[j]) / n
                gradient_memory[j] = new_gradient
                w -= eta * mean_gradient

            elif method == 'Momentum':
                gradient = a @ w - b
                h = 0.9 * h + eta * gradient
                w -= h

            elif method == 'Adam':
                gradient = a @ w - b
                m = 0.9 * m + 0.1 * gradient
                v = 0.999 * v + 0.001 * (gradient ** 2)
                m_hat = m / (1.0 - 0.9 ** (k + 1))
                v_hat = v / (1.0 - 0.999 ** (k + 1))
                w -= eta * m_hat / (np.sqrt(v_hat) + 1e-8)

            else:
                raise ValueError(f'Неизвестный метод: {method}')

            # Если шаг привёл к численной расходимости, этот вариант lambda
            # получает бесконечный validation loss и не может быть выбран лучшим.
            if not np.all(np.isfinite(w)) or np.max(np.abs(w)) > 1e100:
                return w, k + 1, False

            # Одинаковый критерий остановки для всех методов.
            if (k + 1) % check_every == 0:
                current_loss = mse(X_train, y_train, w)
                if not np.isfinite(current_loss):
                    return w, k + 1, False

                if previous_loss is not None and abs(previous_loss - current_loss) <= tol * max(1.0, previous_loss):
                    stable_checks += 1
                else:
                    stable_checks = 0

                previous_loss = current_loss
                if stable_checks >= 5:
                    return w, k + 1, True

    return w, max_iter, True


lambda_grid = np.logspace(-5, 0, 6)
all_results = []
grid_results = {}


def run_experiment(method, use_time_decay, result_name):
    rows = []

    for learning_rate in lambda_grid:
        w, n_iter, valid = fit_optimizer(
            method=method,
            learning_rate=learning_rate,
            use_time_decay=use_time_decay,
        )

        if valid:
            loss_train = mse(X_train, y_train, w)
            loss_val = mse(X_val, y_val, w)
            r2_train = r2_score_manual(X_train, y_train, w)
        else:
            loss_train = np.inf
            loss_val = np.inf
            r2_train = -np.inf

        rows.append({
            'lambda': float(learning_rate),
            'Loss_train': loss_train,
            'R2_train': r2_train,
            'Loss_val': loss_val,
        })

    search_table = pd.DataFrame(rows)
    best_row = search_table.loc[search_table['Loss_val'].idxmin()]
    best_lambda = float(best_row['lambda'])

    # Финально обучаем только на train с lambda, выбранной по validation.
    best_w, n_iter, valid = fit_optimizer(
        method=method,
        learning_rate=best_lambda,
        use_time_decay=use_time_decay,
    )
    if not valid:
        raise RuntimeError('Лучший вариант неожиданно разошёлся при повторном обучении.')

    step_text = (
        f'η(k) = {best_lambda:g}·(1/(1+k))^0.5'
        if use_time_decay
        else f'η = {best_lambda:g}'
    )

    result = {
        'Метод': result_name,
        'Лучший шаг': step_text,
        'lambda': best_lambda,
        'Loss_train': mse(X_train, y_train, best_w),
        'Loss_val': mse(X_val, y_val, best_w),
        'Loss_test': mse(X_test, y_test, best_w),
        'R2_train': r2_score_manual(X_train, y_train, best_w),
        'R2_test': r2_score_manual(X_test, y_test, best_w),
        'Число итераций': int(n_iter),
    }

    all_results.append(result)
    grid_results[result_name] = search_table

    display(pd.DataFrame([result])[[
        'Метод', 'Лучший шаг', 'Loss_train', 'Loss_val',
        'Loss_test', 'R2_train', 'R2_test', 'Число итераций'
    ]].round(6))

    return result


vgd_const = run_experiment('VGD', False, 'VGD, постоянный шаг')


,Метод,Лучший шаг,Loss_train,Loss_val,Loss_test,R2_train,R2_test,Число итераций
0,"VGD, постоянный шаг",η = 0.001,1.192611e+07,2.644347e+07,3.560815e+07,0.799621,0.419406,4560


5. Исследуйте VGD с переменным шагом n(lyamda) по формуле TimeDecayLR (из лекции):

переберите lyamda в логарифмической сетке от 10^-5 до 1

для каждого lyamda:

* обучите VGD на train
* найдите и запомните R^2_train и Loss_train
* протестируйте VGD на val, запомните Loss_val

найдите наилучший lyamda по минимальному Loss_val, запомните его Loss_train, R^2_train, Loss_val;

протестируйте VGD c n(best_lyamda) на test, запомните Loss_test, R^2_test, число итераций на test.

In [5]:
vgd_decay = run_experiment('VGD', True, 'VGD, TimeDecayLR')


,Метод,Лучший шаг,Loss_train,Loss_val,Loss_test,R2_train,R2_test,Число итераций
0,"VGD, TimeDecayLR",η(k) = 0.1·(1/(1+k))^0.5,1.176657e+07,2.656764e+07,3.551304e+07,0.802302,0.420957,2040


6. Исследуйте SGD с постоянным шагом n:

переберите n в логарифмической сетке от 10^-5 до 1

для каждого n:

* обучите SGD на train
* найдите и запомните R^2_train и Loss_train
* протестируйте SGD на val, запомните Loss_val

найдите наилучший n по минимальному Loss_val, запомните его Loss_train, R^2_train, Loss_val;

протестируйте SGD c лучшим n на test, запомните Loss_test, R^2_test, число итераций на test.

In [6]:
sgd_const = run_experiment('SGD', False, 'SGD, постоянный шаг')


,Метод,Лучший шаг,Loss_train,Loss_val,Loss_test,R2_train,R2_test,Число итераций
0,"SGD, постоянный шаг",η = 0.001,1.176723e+07,2.679468e+07,3.560383e+07,0.802291,0.419476,10000


7. Исследуйте SGD с переменным шагом n(lyamda) по формуле TimeDecayLR (из лекции):

переберите lyamda в логарифмической сетке от 10^-5 до 1

для каждого lyamda:

* обучите SGD на train
* найдите и запомните R^2_train и Loss_train
* протестируйте SGD на val, запомните Loss_val

найдите наилучший lyamda по минимальному Loss_val, запомните его Loss_train, R^2_train, Loss_val;

протестируйте SGD c n(best_lyamda) на test, запомните Loss_test, R^2_test, число итераций на test.

In [7]:
sgd_decay = run_experiment('SGD', True, 'SGD, TimeDecayLR')


,Метод,Лучший шаг,Loss_train,Loss_val,Loss_test,R2_train,R2_test,Число итераций
0,"SGD, TimeDecayLR",η(k) = 0.01·(1/(1+k))^0.5,1.234026e+07,2.666127e+07,3.630624e+07,0.792663,0.408024,7600


8. Исследуйте SAG с постоянным шагом n:

переберите n в логарифмической сетке от 10^-5 до 1

для каждого n:

* обучите SAG на train
* найдите и запомните R^2_train и Loss_train
* протестируйте SAG на val, запомните Loss_val

найдите наилучший n по минимальному Loss_val, запомните его Loss_train, R^2_train, Loss_val;

протестируйте SAG c лучшим n на test, запомните Loss_test, R^2_test, число итераций на test.

In [8]:
sag_const = run_experiment('SAG', False, 'SAG, постоянный шаг')


,Метод,Лучший шаг,Loss_train,Loss_val,Loss_test,R2_train,R2_test,Число итераций
0,"SAG, постоянный шаг",η = 0.0001,1.347602e+07,2.909643e+07,3.703384e+07,0.77358,0.39616,10000


9. Исследуйте SAG с переменным шагом n(lyamda) по формуле TimeDecayLR (из лекции):

переберите lyamda в логарифмической сетке от 10^-5 до 1

для каждого lyamda:

* обучите SAG на train
* найдите и запомните R^2_train и Loss_train
* протестируйте SAG на val, запомните Loss_val

найдите наилучший lyamda по минимальному Loss_val, запомните его Loss_train, R^2_train, Loss_val;

протестируйте SAG c n(best_lyamda) на test, запомните Loss_test, R^2_test, число итераций на test.

In [9]:
sag_decay = run_experiment('SAG', True, 'SAG, TimeDecayLR')


,Метод,Лучший шаг,Loss_train,Loss_val,Loss_test,R2_train,R2_test,Число итераций
0,"SAG, TimeDecayLR",η(k) = 0.01·(1/(1+k))^0.5,1.242175e+07,2.706280e+07,3.797310e+07,0.791294,0.380845,10000


10. Исследуйте Momentum с постоянным шагом n:

переберите n в логарифмической сетке от 10^-5 до 1

для каждого n:

* обучите Momentum на train
* найдите и запомните R^2_train и Loss_train
* протестируйте Momentum на val, запомните Loss_val

найдите наилучший n по минимальному Loss_val, запомните его Loss_train, R^2_train, Loss_val;

протестируйте Momentum c лучшим n на test, запомните Loss_test, R^2_test, число итераций на test.

In [10]:
momentum_const = run_experiment('Momentum', False, 'Momentum, постоянный шаг')


,Метод,Лучший шаг,Loss_train,Loss_val,Loss_test,R2_train,R2_test,Число итераций
0,"Momentum, постоянный шаг",η = 0.0001,1.192626e+07,2.644330e+07,3.560694e+07,0.799619,0.419426,4560


11. Исследуйте Momentum с переменным шагом n(lyamda) по формуле TimeDecayLR (из лекции):

переберите lyamda в логарифмической сетке от 10^-5 до 1

для каждого lyamda:

* обучите Momentum на train
* найдите и запомните R^2_train и Loss_train
* протестируйте Momentum на val, запомните Loss_val

найдите наилучший lyamda по минимальному Loss_val, запомните его Loss_train, R^2_train, Loss_val;

протестируйте Momentum c n(best_lyamda) на test, запомните Loss_test, R^2_test, число итераций на test.

In [11]:
momentum_decay = run_experiment('Momentum', True, 'Momentum, TimeDecayLR')


,Метод,Лучший шаг,Loss_train,Loss_val,Loss_test,R2_train,R2_test,Число итераций
0,"Momentum, TimeDecayLR",η(k) = 0.01·(1/(1+k))^0.5,1.176614e+07,2.656388e+07,3.551101e+07,0.802309,0.42099,2040


12. Исследуйте Adam с постоянным шагом n:

переберите n в логарифмической сетке от 10^-5 до 1

для каждого n:

* обучите Adam на train
* найдите и запомните R^2_train и Loss_train
* протестируйте Adam на val, запомните Loss_val

найдите наилучший n по минимальному Loss_val, запомните его Loss_train, R^2_train, Loss_val;

протестируйте Adam c лучшим n на test, запомните Loss_test, R^2_test, число итераций на test.

In [12]:
adam_const = run_experiment('Adam', False, 'Adam, постоянный шаг')


,Метод,Лучший шаг,Loss_train,Loss_val,Loss_test,R2_train,R2_test,Число итераций
0,"Adam, постоянный шаг",η = 1,1.142014e+07,2.891033e+07,3.739823e+07,0.808122,0.390219,9640


13. Исследуйте Adam с переменным шагом n(lyamda) по формуле TimeDecayLR (из лекции):

переберите lyamda в логарифмической сетке от 10^-5 до 1

для каждого lyamda:

* обучите Adam на train
* найдите и запомните R^2_train и Loss_train
* протестируйте Adam на val, запомните Loss_val

найдите наилучший lyamda по минимальному Loss_val, запомните его Loss_train, R^2_train, Loss_val;

протестируйте Adam c n(best_lyamda) на test, запомните Loss_test, R^2_test, число итераций на test.

In [13]:
adam_decay = run_experiment('Adam', True, 'Adam, TimeDecayLR')


,Метод,Лучший шаг,Loss_train,Loss_val,Loss_test,R2_train,R2_test,Число итераций
0,"Adam, TimeDecayLR",η(k) = 1·(1/(1+k))^0.5,8.250150e+07,1.108425e+08,8.396279e+07,-0.386164,-0.36902,10000


14. Постройте итоговую сравнительную таблицу со следующими столбцами:

1) название метода

2) значение лучшего шага (если n) или функция лучшего шага (если n(lyamda))

3) Loss_train

4) Loss_test

5) R^2 train

6) R^2 test

7) число итераций на test

In [14]:
comparison_table = pd.DataFrame(all_results)[[
    'Метод', 'Лучший шаг', 'Loss_train', 'Loss_test',
    'R2_train', 'R2_test', 'Число итераций'
]].copy()

comparison_table.round({
    'Loss_train': 4,
    'Loss_test': 4,
    'R2_train': 4,
    'R2_test': 4,
})


,Метод,Лучший шаг,Loss_train,Loss_test,R2_train,R2_test,Число итераций
0,"VGD, постоянный шаг",η = 0.001,1.192611e+07,3.560815e+07,0.7996,0.4194,4560
1,"VGD, TimeDecayLR",η(k) = 0.1·(1/(1+k))^0.5,1.176657e+07,3.551304e+07,0.8023,0.4210,2040
2,"SGD, постоянный шаг",η = 0.001,1.176723e+07,3.560383e+07,0.8023,0.4195,10000
3,"SGD, TimeDecayLR",η(k) = 0.01·(1/(1+k))^0.5,1.234026e+07,3.630624e+07,0.7927,0.4080,7600
4,"SAG, постоянный шаг",η = 0.0001,1.347602e+07,3.703384e+07,0.7736,0.3962,10000
5,"SAG, TimeDecayLR",η(k) = 0.01·(1/(1+k))^0.5,1.242175e+07,3.797310e+07,0.7913,0.3808,10000
6,"Momentum, постоянный шаг",η = 0.0001,1.192626e+07,3.560694e+07,0.7996,0.4194,4560
7,"Momentum, TimeDecayLR",η(k) = 0.01·(1/(1+k))^0.5,1.176614e+07,3.551101e+07,0.8023,0.4210,2040
8,"Adam, постоянный шаг",η = 1,1.142014e+07,3.739823e+07,0.8081,0.3902,9640
9,"Adam, TimeDecayLR",η(k) = 1·(1/(1+k))^0.5,8.250150e+07,8.396279e+07,-0.3862,-0.3690,10000


15. Сделайте вывод о том, какой метод и шаг линейной регрессии самый лучший для данной выборки и ответьте на вопросы:

1) почему именно этот метод и этот шаг самый лучший (по каким данным из таблицы вы сделали такой вывод)

2) расскажите простыми словами суть R^2_train и R^2_test?

3) как R^2_train и R^2_test помогают сравнивать методы? почему оба эти значения надо вычислять для данного выбора?

### Вывод

1. **Лучший вариант для этой выборки — Momentum с TimeDecayLR при $\lambda=0.01$**, то есть
   $\eta_k=0.01\left(\frac{1}{1+k}\right)^{0.5}$.
   После выбора $\lambda$ только по `validation` этот вариант дал на `test` `Loss_test ≈ 35 511 009` и `R²_test ≈ 0.421`. Это минимальный `Loss_test` и максимальный `R²_test` в итоговой таблице. Он также остановился за 2040 итераций. Результат VGD с TimeDecayLR почти такой же, поэтому разница между ними небольшая.

2. **R²_train** показывает, какую долю разброса целевой переменной модель объясняет на обучающей выборке. **R²_test** показывает то же самое на новых данных, которые не участвовали в обучении. Чем R² ближе к 1, тем лучше модель объясняет данные; значение около 0 означает качество примерно на уровне предсказания среднего значения, отрицательное — хуже такого базового предсказания.

3. `R²_train` нужен, чтобы понимать, насколько хорошо модель подогналась под обучающие данные, а `R²_test` — насколько это качество переносится на новые данные. Если `R²_train` высокий, а `R²_test` заметно ниже, модель хуже обобщает на новых объектах. Поэтому при сравнении методов нужно смотреть на оба значения вместе с `Loss_test`.
